### ISSUE TITLE
"This application cannot tree shake icons fonts" error occurs after upgrading flutter to 1.22.0, breaking build

### ISSUE THREAD

**Issue by sceee** [ziofat/material_design_icons_flutter#27]
After upgrading flutter to 1.22.0, the following error occurred when trying to build the Android App bundle
```
This application cannot tree shake icons fonts. It has non-constant instances of IconData at the following locations:
  - file:///C:/Dev/flutter/.pub-cache/hosted/pub.dartlang.org/material_design_icons_flutter-4.0.5655/lib/material_design_icons_flutter.dart:5680:12
                                                                        
                                                                        
FAILURE: Build failed with an exception.

* Where:
Script 'C:\Dev\flutter\packages\flutter_tools\gradle\flutter.gradle' line: 904

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command 'C:\Dev\flutter\bin\flutter.bat'' finished with non-zero exit value 1
```

**Comment by TarekkMA** [comment]
Build with ``--no-tree-shake-icons`` to disable this feature.

**Comment by sceee** [comment]
> Build with `--no-tree-shake-icons` to disable this feature.

Since it only appeared in Flutter 1.22.0 after upgrading from 1.20.4, I don't think this is the appropriate solution, isn't it? Also, how would I specify this for an iOS build using XCode?

**Comment by lvlrSajjad** [comment]
I have the same problem using this command:
```
flutter build ios
```

When I use flutter run --release
It works fine ...

**Comment by sceee** [comment]
@lvlrSajjad when using `flutter build ios --no-tree-shake-icons` it should work (also archiving it afterwards using xcode). This seems to be the only solution right now.

**Comment by ziofat** [comment]
I think this might because of we have a fromString method to generate IconData, by using this method we will lost all the benefits of tree-shaking. I will look into this to find a solution. PR is welcome also.

**Comment by sceee** [comment]
@ziofat yes, that's the reason as far as I understand it. According to what I read I don't know if there is a solution at all as what I understood is:
as soon as something with _any_ icon font is done using the `fromString`, flutter can no longer safely detect which icons are used and tree shake off the unused ones - which makes sense as in my case, they are used dynamically and must not be stripped off (otherwise dynamically loading the icons by string would fail for stripped off icons).

**Comment by lvlrSajjad** [comment]
> I think this might because of we have a fromString method to generate IconData, by using this method we will lost all the benefits of tree-shaking. I will look into this to find a solution. PR is welcome also.

So if I replace all fromString methods with the other method the problem will be solved In my app?

**Comment by ziofat** [comment]
@lvlrSajjad I cannot guarantee that cuz I am temporally lost access to my computer. But I do have strong confidence of this.

**Comment by lvlrSajjad** [comment]
@ziofat It worked for me. :-D At least for now.
@sceee Just replace fromString icons with legit ones and if you have some function with dynamic icons use a function like this
```
IconData _stringToIcon(String input) {
    if(input == "cancel") {
      return MdiIcons.cancel;
    } else if (input == "phone") {
      return MdiIcons.phone;
    } else if (input == "chat") {
      return MdiIcons.chat;
    } else if (input == "thumbUp") {
      return MdiIcons.thumbUp;
    } else if (input == "infinity") {
      return MdiIcons.infinity;
    } else if (input == "redoVariant") {
      return MdiIcons.redoVariant;
    } else if (input == "key") {
      return MdiIcons.key;
    } else if (input == "eye") {
      return MdiIcons.eye;
    } else if (input == "accountSearch") {
      return MdiIcons.accountSearch;
    } else if (input == "eyeOff") {
      return MdiIcons.eyeOff;
    } else if (input == "reload") {
      return MdiIcons.reload;
    } else if (input == "clockOutline") {
      return MdiIcons.clockOutline;
    }
  }
```

**Comment by ziofat** [comment]
@lvlrSajjad Thank you for testing this.
As this is an expecting behaviour and has already been mentioned in readme, I will close this issue after I add some more description in readme file.

**Comment by bambinoua** [comment]
https://github.com/ziofat/material_design_icons_flutter/issues/54#issuecomment-1550022999
